extensions/amp-ad/0.1/amp-ad.js

```diff
@@ -89,318 +60,35 @@ class AmpAd extends AMP.BaseElement {
-    if (this.intersectionObserver_) {
-      this.intersectionObserver_.onViewportCallback(inViewport);
-    }
-    this.sendEmbedInfo_(inViewport);
-  }
-
-  /**
-   * @param {boolean} inViewport
-   * @private
-   */
-  sendEmbedInfo_(inViewport) {
-    if (this.iframe_) {
-      const targetOrigin =
-          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
-      postMessage(this.iframe_, 'embed-state', {
-        inViewport,
-        pageHidden: !this.viewer_.isVisible(),
-      }, targetOrigin, /* opt_is3P */ true);
-    }
-  }
-
-  /** @override  */
-  overflowCallback(overflown, requestedHeight, requestedWidth) {
-    if (overflown) {
-      const targetOrigin =
-          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
-      postMessage(
-          this.iframe_,
-          'embed-size-denied',
-          {requestedHeight, requestedWidth},
-          targetOrigin,
-          /* opt_is3P */ true);
-    }
-  }
-
-  /**
-   * Updates the element's dimensions to accommodate the iframe's
-   *    requested dimensions.
-   * @param {number|undefined} newWidth
-   * @param {number|undefined} newHeight
-   * @private
-   */
-  updateSize_(newHeight, newWidth) {
-    this.attemptChangeSize(newHeight, newWidth, () => {
-      const targetOrigin =
-          this.iframe_.src ? parseUrl(this.iframe_.src).origin : '*';
-      postMessage(
-          this.iframe_,
-          'embed-size-changed',
-          {requestedHeight: newHeight, requestedWidth: newWidth},
-          targetOrigin,
-          /* opt_is3P */ true);
-    });
-  }
-
-  /**
-   * Activates the fallback if the ad reports that the ad slot cannot
-   * be filled.
-   * @private
-   */
-  noContentHandler_() {
-    // If a fallback does not exist attempt to collapse the ad.
-    if (!this.fallback_) {
-      this.attemptChangeHeight(0, () => {
-        this.element.style.display = 'none';
-      });
-    }
-    this.deferMutate(() => {
-      if (this.fallback_) {
-        // Hide placeholder when falling back.
-        if (this.placeholder_) {
-          this.togglePlaceholder(false);
-        }
-        this.toggleFallback(true);
-      }
-      // Remove the iframe only if it is not the master.
-      if (this.iframe_.name.indexOf('_master') == -1) {
-        removeElement(this.iframe_);
-        this.iframe_ = null;
-      }
-    });
+    if (!type) {
+      // Unspecified or empty type.  Nothing to do here except bail out.
+      return;
+    }
+    // TODO(tdrl): Check amp-ad registry to see if they have this already.
+    if (!a4aRegistry[type] ||
+        !a4aRegistry[type](this.getWin(), this.element)) {
+      // Network either has not provided any A4A implementation or the
+      // implementation exists, but has explicitly chosen not to handle this
+      // tag as A4A.  Fall back to the 3p implementation.
+      const newChild = this.element.ownerDocument.createElement(TAG_3P_IMPL);
+      copyAttributes_(this.element, newChild);
```

Comment:
The 3p extension code is but not the network specific which we do not want to bundle as hopefully there will be many in the future.  Given the non-3p will require the child attach method, we elected to do the same for 3p for simplicity.  @tdrl found in testing that calling the 3p constructor directly required replaying any lifecycle events for the new instance while appending a new element does not.